crates/bevy_render/src/batching/gpu_preprocessing.rs

```diff
@@ -0,0 +1,313 @@
+//! Batching functionality when GPU preprocessing is in use.
+
+use std::marker::PhantomData;
+
+use bevy_ecs::{
+    entity::Entity,
+    query::With,
+    system::{Query, Res, ResMut, Resource, StaticSystemParam},
+};
+use bevy_encase_derive::ShaderType;
+use bevy_utils::EntityHashMap;
+use bytemuck::{Pod, Zeroable};
+use smallvec::smallvec;
+use wgpu::{BindingResource, BufferUsages};
+
+use crate::{
+    render_phase::{
+        BinnedPhaseItem, BinnedRenderPhase, BinnedRenderPhaseBatch, CachedRenderPipelinePhaseItem,
+        SortedPhaseItem, SortedRenderPhase,
+    },
+    render_resource::{BufferVec, GpuArrayBufferIndex, GpuArrayBufferable, UninitBufferVec},
+    renderer::{RenderDevice, RenderQueue},
+    view::ViewTarget,
+};
+
+use super::{BatchMeta, GetFullBatchData};
+
+/// The GPU buffers holding the data needed to render batches.
+///
+/// For example, in the 3D PBR pipeline this holds `MeshUniform`s, which are the
+/// `BD` type parameter in that mode.
+///
+/// We have a separate *buffer data input* type (`BDI`) here, which a compute
+/// shader is expected to expand to the full buffer data (`BD`) type. GPU
+/// uniform building is generally faster and uses less GPU bus bandwidth, but
+/// only implemented for some pipelines (for example, not in the 2D pipeline at
+/// present) and only when compute shader is available.
+#[derive(Resource)]
+pub struct BatchedInstanceBuffers<BD, BDI>
+where
+    BD: GpuArrayBufferable + Sync + Send + 'static,
+    BDI: Pod,
+{
+    /// A storage area for the buffer data that the GPU compute shader is
+    /// expected to write to.
+    ///
+    /// There will be one entry for each index.
+    pub data_buffer: UninitBufferVec<BD>,
+
+    /// The index of the buffer data in the current input buffer that
+    /// corresponds to each instance.
+    ///
+    /// This is keyed off each view. Each view has a separate buffer.
+    pub work_item_buffers: EntityHashMap<Entity, BufferVec<PreprocessWorkItem>>,
+
+    /// The uniform data inputs for the current frame.
+    ///
+    /// These are uploaded during the extraction phase.
+    pub current_input_buffer: BufferVec<BDI>,
+
+    /// The uniform data inputs for the previous frame.
+    ///
+    /// The indices don't generally line up between `current_input_buffer`
+    /// and `previous_input_buffer`, because, among other reasons, entities
+    /// can spawn or despawn between frames. Instead, each current buffer
+    /// data input uniform is expected to contain the index of the
+    /// corresponding buffer data input uniform in this list.
+    pub previous_input_buffer: BufferVec<BDI>,
+}
+
+/// One invocation of the preprocessing shader: i.e. one mesh instance in a
+/// view.
+#[derive(Clone, Copy, Pod, Zeroable, ShaderType)]
```

Comment:
Why `ShaderType`? Seems unnecessary?

Reply:
`bind_group_layout_entries::binding_types::storage_buffer_read_only` requires that its type parameter be `ShaderType`.